File: 19-graphs1/4-dfs/main.py
# Depth-First Search (DFS) - Python Example

def dfs(graph, node, visited=None):
    if visited is None:
        visited = set()
    visited.add(node)
    print(node, end=" ")
    for neighbor in graph.get(node, []):
        if neighbor not in visited:
            dfs(graph, neighbor, visited)

# Example graph
graph = {
    "A": ["B", "C"],
    "B": ["D"],
    "C": ["D"],
    "D": []
}

print("DFS traversal starting from A:")
dfs(graph, "A")

File: 19-graphs1/4-dfs/main.cpp
// Depth-First Search (DFS) - C++ Example

#include <iostream>
#include <unordered_map>
#include <vector>
#include <unordered_set>

void dfs(const std::unordered_map<char, std::vector<char>>& graph, char node, std::unordered_set<char>& visited) {
    visited.insert(node);
    std::cout << node << " ";
    for (char neighbor : graph.at(node)) {
        if (visited.find(neighbor) == visited.end()) {
            dfs(graph, neighbor, visited);
        }
    }
}

int main() {
    std::unordered_map<char, std::vector<char>> graph = {
        {'A', {'B', 'C'}},
        {'B', {'D'}},
        {'C', {'D'}},
        {'D', {}}
    };

    std::unordered_set<char> visited;
    std::cout << "DFS traversal starting from A: ";
    dfs(graph, 'A', visited);
    std::cout << std::endl;

    return 0;
}
